### PR TITLE
ISPN-11838 CLI with Kubernetes Operator superpowers

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-11838

The native bits go to a separate infinispan-quarkus PR

https://github.com/infinispan/infinispan-designs/pull/20

### CODE REVIEW

File: cli/cli-client/src/main/resources/operator/rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (dmvolod) [inline cli/cli-client/src/main/resources/operator/rbac.yaml:1]: @tristantarrant please pay attention, that Role now is under rights assessment
https://github.com/infinispan/infinispan-operator/pull/537

File: cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Install.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+   }
+
+   @Override
+   public CommandResult exec(ContextAwareCommandInvocation invocation) {
+      KubernetesClient kubernetesClient = invocation.getContext().getKubernetesClient();
+
+      try(InputStream crd = this.getClass().getClassLoader().getResourceAsStream("operator/rbac.yaml")) {
+         kubernetesClient.load(crd).inNamespace(namespace).createOrReplace();
+      } catch (IOException e) {
+         throw Messages.MSG.genericError("rbac.yaml", e);
```
Comment (dmvolod) [inline cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Install.java:44]: @tristantarrant please pay attention, that ClusterRole is also requires.
https://github.com/infinispan/infinispan-operator/blob/master/deploy/clusterrole.yaml
Role and ClusterRole are sometimes changing and need method for sync between operator and this repos

File: cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Kube.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            Interactive.class,
+            Create.class,
+            Ls.class,
+            Remove.class,
+            Version.class
+      })
+public class Kube implements Command {
+   public static final String CMD = "kube";
+   public static final String DEFAULT_NAMESPACE = "default";
+   public static final String DEFAULT_SERVICE_NAME = "infinispan";
```
Comment (dmvolod) [inline cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Kube.java:38]: Service is a bit overload. Kubernetes Service and Infinispan service are confusing. Maybe call `DEFAULT_INFINISPAN_NAME` or `DEFAULT_CLUSTER_NAME`?

**Comment by tristantarrant** [comment]
Thanks @dmvolod : I will be updating the Kubernetes code against the latest version of the operator
These are still bits that I had lying around a few months back.

**Comment by tristantarrant** [comment]
Ok, this is now ready for review.

**Comment by tristantarrant** [comment]
Btw, in the end I didn't bother with writing tests for the operator part because it would just mean mocking everything and I'd rather spend time on real integration tests.

**Comment by ryanemerson** [comment]
CI is failing due to checkstyle

**Review by ryanemerson — CHANGES_REQUESTED** [review]
The help `.adoc` and docs are currently missing, is this coming in another PR?

I tried executing:

```
java -jar -Dinfinispan.cli.kubernetes=true  target/infinispan-cli-client-12.0.0-SNAPSHOT.jar install -n cli
```
I also tried with the long form "--namespace" and I get:

```
KubernetesClientException: Namespace not specified. But operation requires namespace.
```

File: cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Uninstall.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+   @Override
+   public CommandResult exec(ContextAwareCommandInvocation invocation) {
+      KubernetesClient client = ((KubernetesContextImpl)invocation.getContext()).getKubernetesClient();
+      String crd = Kube.loadResourceAsString("operator/operator-install.yaml", new TypedProperties().setProperty("NAMESPACE", namespace));
+
+      ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable<HasMetadata, Boolean> op = client.load(new ByteArrayInputStream(crd.getBytes(StandardCharsets.UTF_8)));
+      if (namespace != null) {
+         op.inNamespace(namespace);
+      }
+      op.createOrReplace();
```
Comment (dmvolod) [inline cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Uninstall.java:49]: Seems should be delete() not createOrReplace() for uninstall

File: cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Install.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+   @Override
+   public CommandResult exec(ContextAwareCommandInvocation invocation) {
+      KubernetesClient client = ((KubernetesContextImpl)invocation.getContext()).getKubernetesClient();
+      String crd = Kube.loadResourceAsString("operator/operator-install.yaml", new TypedProperties().setProperty("NAMESPACE", namespace));
+
+      ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable<HasMetadata, Boolean> op = client.load(new ByteArrayInputStream(crd.getBytes(StandardCharsets.UTF_8)));
+      if (namespace != null) {
+         op.inNamespace(namespace);
+      }
+      op.createOrReplace();
```
Comment (dmvolod) [inline cli/cli-client/src/main/java/org/infinispan/cli/commands/kubernetes/Install.java:49]: Install has following cases need to be handled
- The operator can be already installed with other or the same version. CRD and ClusterRoles are cluster wide resources and can be detected before install
- The installation account may have insufficient rights and cluster may remain some garbage resources after unsuccessful installation

**Comment by ryanemerson** [comment]
I'm getting the following errors:

# Install
Cmd: `java -jar -Dinfinispan.cli.kubernetes=true  target/infinispan-cli-client-12.0.0-SNAPSHOT.jar install -n cli`
Error:
```
JsonMappingException: No resource type found for:operators.coreos.com/v1alpha2#OperatorGroup
 at [Source: (BufferedInputStream); line: 9, column: 1]
```

# Uninstall
Cmd: `java -jar -Dinfinispan.cli.kubernetes=true  target/infinispan-cli-client-12.0.0-SNAPSHOT.jar uninstall -n cli`
Error: `NullPointerException: null

The operator is not removed.


# Create cluster
The operator is not able to provision the cluster due to the following error in the operator logs:

```
{"level":"error","ts":1603974237.6113305,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"infinispan-controller","request":"cli/infinispan","error":"Infinispan.infinispan.org \"infinispan\" is invalid: status.statefulSetName: Required value","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"}
 remerson@localhost  ~/workspace/RedHat/infinispan/infinispan-alt/cli   tristantarr
```

I think updating `operator-install.yaml` to the [latest version](https://github.com/infinispan/infinispan-operator/blob/master/deploy/operator-install.yaml) should fix this.

**Comment by tristantarrant** [comment]
@ryanemerson I've fixed the install/uninstall issues.
Uninstall currently just removes the subscription, but we may want to do a bit more.

**Comment by ryanemerson** [comment]
@tristantarrant Create and delete work no problem, however install/uninstall I always get a 404 error even though the operation works.

Install:
```
KubernetesClientException: Failure executing: POST at: https://192.168.39.11:8443/apis/operators.coreos.com/v1/namespaces/cli2/operatorgroups. Message: 404 page not found
```

Uninstall:
```
KubernetesClientException: Failure executing: DELETE at: https://192.168.39.11:8443/apis/operators.coreos.com/v1alpha1/namespaces/cli2/subscriptions/infinispan. Message: 404 page not found
```

I'm using minikube to test, what env are you using? I wonder if that's why you aren't seeing these errors.

**Comment by tristantarrant** [comment]
Did you install the OLM?
I'm using Kind: https://github.com/tristantarrant/infinispan-with-kind

**Comment by ryanemerson** [comment]
No, that was the issue :sweat_smile: 

Thanks @tristantarrant

**Pull request merged**